// @privy-io/privy-node/dist/http.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Http = void 0;
const axios_1 = __importDefault(require("axios"));
const errors_1 = require("./errors");
class Http {
    constructor(authenticate, defaults) {
        this.authenticate = authenticate;
        this.defaults = defaults;
    }
    async get(path, config) {
        try {
            return await axios_1.default.get(path, await this.buildConfig(config));
        }
        catch (e) {
            throw (0, errors_1.wrapApiError)(e);
        }
    }
    async post(path, data, config) {
        try {
            return await axios_1.default.post(path, data, await this.buildConfig(config));
        }
        catch (e) {
            throw (0, errors_1.wrapApiError)(e);
        }
    }
    async delete(path, config) {
        try {
            return await axios_1.default.delete(path, await this.buildConfig(config));
        }
        catch (e) {
            throw (0, errors_1.wrapApiError)(e);
        }
    }
    async buildConfig(config) {
        config = config || {};
        config.headers = config.headers || {};
        if (this.authenticate) {
            const token = await this.authenticate();
            config.headers.authorization = `Bearer ${token}`;
        }
        return { ...this.defaults, ...config };
    }
}
exports.Http = Http;
