// @privy-io/privy-node/dist/paths.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.siwePath = exports.siweNoncePath = exports.integrityHashPath = exports.fileDownloadsPath = exports.fileUploadsPath = exports.wrapperKeyPath = exports.batchDataKeyPath = exports.dataKeyPath = exports.batchUserDataPath = exports.userDataPath = void 0;
const userDataPath = (userId, fields) => {
    const path = `/users/${userId}/data`;
    const query = ['new=t'];
    if (fields && fields.length > 0) {
        const uriEncodedFields = fields.map(encodeURIComponent);
        query.push(`fields=${uriEncodedFields.join(',')}`);
    }
    return `${path}?${query.join('&')}`;
};
exports.userDataPath = userDataPath;
const batchUserDataPath = (fields, options) => {
    const path = `/data`;
    const query = [];
    if (options.cursor) {
        query.push(`cursor=${options.cursor}`);
    }
    if (options.limit) {
        query.push(`limit=${options.limit}`);
    }
    if (fields && fields.length > 0) {
        const uriEncodedFields = fields.map(encodeURIComponent);
        query.push(`fields=${uriEncodedFields.join(',')}`);
    }
    return `${path}?${query.join('&')}`;
};
exports.batchUserDataPath = batchUserDataPath;
const dataKeyPath = (userId) => {
    return `/key_manager/users/${userId}/data_key`;
};
exports.dataKeyPath = dataKeyPath;
const batchDataKeyPath = () => {
    return `/key_manager/data_key`;
};
exports.batchDataKeyPath = batchDataKeyPath;
const wrapperKeyPath = (userId) => {
    return `/key_manager/users/${userId}/wrapper_key`;
};
exports.wrapperKeyPath = wrapperKeyPath;
const fileUploadsPath = (userId, fieldId) => {
    return `/users/${userId}/fields/${fieldId}/files`;
};
exports.fileUploadsPath = fileUploadsPath;
const fileDownloadsPath = (userId, fieldId, fileId) => {
    return `/users/${userId}/fields/${fieldId}/files/${fileId}/contents`;
};
exports.fileDownloadsPath = fileDownloadsPath;
const integrityHashPath = (integrityHash) => {
    return `/data/${integrityHash}?new=t`;
};
exports.integrityHashPath = integrityHashPath;
const siweNoncePath = () => {
    return '/auth/siwe/nonce';
};
exports.siweNoncePath = siweNoncePath;
const siwePath = () => {
    return '/auth/siwe';
};
exports.siwePath = siwePath;
