// @privy-io/crypto/dist/crypto/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setCrypto = void 0;
/**
 * Creates a proxy to an underlying implementation of crypto operations.
 *
 * This is used to create a stable object by which callers can operate
 * against while allowing the underlying implementation to be configured.
 * It is used to configure the implementation depending on whether the
 * environment is node or the browser.
 *
 * @internal
 */
function createCryptoProxy() {
    var crypto = undefined;
    function setCrypto(impl) {
        crypto = impl;
    }
    var proxy = {
        get csprng() {
            return crypto.csprng;
        },
        get sha256() {
            return crypto.sha256;
        },
        get aesGCMEncrypt() {
            return crypto.aesGCMEncrypt;
        },
        get aesGCMDecrypt() {
            return crypto.aesGCMDecrypt;
        },
        get aesGCMEncryptionKey() {
            return crypto.aesGCMEncryptionKey;
        },
        get rsaOAEPEncrypt() {
            return crypto.rsaOAEPEncrypt;
        },
    };
    return [proxy, setCrypto];
}
var _a = createCryptoProxy(), crypto = _a[0], _setCrypto = _a[1];
exports.default = crypto;
exports.setCrypto = _setCrypto;
